src/preamble.js

```diff
@@ -496,6 +495,11 @@ function UTF8ArrayToString(u8Array, idx) {
 // a copy of that string as a Javascript String object.
 
 function UTF8ToString(ptr) {
+#if ASSERTIONS
+  // Require that first element of the heap is \0 so that it terminates null strings without having to run an explicit check for that
+  // in UTF8ArrayToString(). (If UTF8ArrayToString() is called on arrays other than the heap, index 0 in them might not have this property)
+  assert(HEAPU8[0] == 0);
```

Comment:
this seems pretty risky to me. there might be code that uses 0 as a temporary location for copying stuff, for example - in fact, I believe binaryen's wasm2js currently does (which is not great).

what is the benefit of assuming this is zero? i'm not sure where in this PR we take advantage of that.

Reply:
Only the TEXTDECODER==2 path avoids assuming HEAP8[0]==0 by the virtue of that `endPtr>idx` check that was added. The other paths do still scan through address zero for a string there, and without this code, there was a bug that `UTF8ToString(0)` was returning `"emsc"`.

If we don't do this assumption, then in TEXTDECODER<2 paths we'll have to add a `if (!ptr) return '';` check, which I'd rather not do ; not only because it adds code size, but because it also adds an extra if(). With the assumption of HEAP8[0]==0, TEXTDECODER<2 path will automatically be temp garbage free on empty strings.

Reply:
On second thought, actually, let me back off from this scheme. I realize that it will not be very resilient, because if any other code accidentally writes to the null pointer in release build, then all strings will begin to convert to JS strings as random data. That would of course be ok from C undefined behavior perspective, but it does not actually cost any more code bytes to hoist the check out from TEXTDECODER==2 path, and we will still be garbage free. Reverted back to old form, and added a regular if check to UTF8ToString().